### PR TITLE
Taxometric hyperparams

### PR DESCRIPTION
**Pull request by tmckeown311**
# Overview

**Type of Change:** New Feature/Debug

**Summary:** Fixed bug where training transformation was not saved or used for evaluation with the genus specific models. Added hyperparameter tuning to the genus specific trainer and added an option to the simulator to run hyperparameter tuning instead of training.

## UI/UX Implementation

None

## Model Updates

None

### Data Models

None

### Object-Oriented Models

Added another transformation that gets saved when training genus specific models

## End-to-End Testing Instructions

First try and retrain the models so that the transformation file is saved and then check that the evaluator works as intended. The evaluator should have a higher accuracy than it previously had. Then test that the hyperparameter option in the simulator works as intended and runs hyper tuning.

### CODE REVIEW

File: src/alt_training_program.py [hunk 1]
```diff
@@ -535,7 +535,7 @@ def save_transformation(self, transformation, angle):
                 dill.dump(transformation, f)
 
         elif angle == 1:
-            with open(globals.all_transformations, "wb") as f:
```
Comment (JoeCx) [inline src/alt_training_program.py:538]: Good fix

File: src/genus_specific_model_trainer.py [hunk 2]
```diff
@@ -3,32 +3,39 @@
...
 from torch.utils.data import DataLoader, Dataset
 from torchvision import transforms, models
 import torch
 import dill
 from sklearn.model_selection import train_test_split
+from sklearn.model_selection import StratifiedKFold
 from sklearn.metrics import f1_score
+import optuna
 from transformation_classes import HistogramEqualization
+from data_augmenter import DataAugmenter
```
Comment (JoeCx) [inline src/genus_specific_model_trainer.py:20]: For data leakage fix

File: src/genus_specific_model_trainer.py [hunk 3]
```diff
@@ -67,17 +76,91 @@ def load_model(self, num_classes):
...
         labels = [class_string_dictionary[label] for label in classes]
-        train_x, test_x, train_y, test_y = train_test_split(
-            image_binaries, labels, test_size=0.2, random_state=42
+
+        # Split by index for safe DataFrame reconstruction
+        indices = np.arange(len(dataframe))
+        train_idx, test_idx = train_test_split(
+            indices, test_size=0.2, stratify=labels, random_state=42
         )
+        # Create train/test DataFrames
```
Comment (JoeCx) [inline src/genus_specific_model_trainer.py:145]: Data augmentation fixes

File: src/genus_specific_model_trainer.py [hunk 4]
```diff
@@ -111,7 +194,15 @@ def train_genus(self, genus, num_epochs):
...
-        train_dataset = ImageDataset(train_x, train_y, transform=self.transformation)
+        # Define image training transformations, placeholder for preprocessing
+        self.train_transformations = self.create_train_transformations(
+            rotation_degree=5,
+            brightness=0.1,
+            contrast=0.1,
+            erasing=(0.5, (0.02, 0.15))
+        )
+
+        train_dataset = ImageDataset(train_x, train_y, transform=self.train_transformations)
```
Comment (JoeCx) [inline src/genus_specific_model_trainer.py:205]: Good change here

File: src/genus_specific_model_trainer.py [hunk 5]
```diff
@@ -184,6 +275,181 @@ def training_evaluation(self, num_epochs, train_loader, test_loader, model, genu
             self.model_accuracies[genus] = best_macro_f1
             print(f"Best Macro F1: {100 * best_macro_f1:.2f}% - model loaded.")
 
+
+    def gs_hyperparameter_training_evaluation(self, num_epochs, train_loader, test_loader, input_model, lr):
```
Comment (JoeCx) [inline src/genus_specific_model_trainer.py:279]: Good job adding hyperparameter tuning here

File: src/genus_specific_simulator.py [hunk 6]
```diff
@@ -25,6 +25,17 @@
             break
         print("Invalid Input. Please enter 1 or 2.")
 
+    while True:
```
Comment (JoeCx) [inline src/genus_specific_simulator.py:28]: Nice updates to simulator file

File: src/genus_specific_simulator.py [hunk 7]
```diff
@@ -39,25 +50,17 @@
...
-        print(f"FRON: {(df['View'] == 'FRON').sum()}")
-        print(f"LATE: {(df['View'] == 'LATE').sum()}")
-
     genus_list = df['Genus'].unique().tolist()
-
-    print(f"Training for each genus in this list: {genus_list}")
-    genus_specific_tp = GenusSpecificModelTrainer(df)
-    for genus in genus_list:
-        genus_specific_tp.train_genus(genus, 20)
+    genus_specific_tp = GenusSpecificModelTrainer(dataframe=df, augment=augment)
```
Comment (JoeCx) [inline src/genus_specific_simulator.py:54]: Data augmentation fix

File: testing/test_genus_specific_model_trainer.py [hunk 8]
```diff
@@ -26,13 +26,20 @@ def create_mock_image_blob():
...
         # Create a mock DataFrame for testing
         self.mock_dataframe = pd.DataFrame({
-            "Genus": ["GenusA", "GenusB", "GenusC", "GenusD", "GenusE",
-                      "GenusF", "GenusG", "GenusH", "GenusI", "GenusJ"],
-            "Species": ["SpeciesA", "SpeciesB", "SpeciesC", "SpeciesD", "SpeciesE",
-                        "SpeciesF", "SpeciesG", "SpeciesH", "SpeciesI", "SpeciesJ"],
-            "UniqueID": ["ID1", "ID2", "ID3", "ID4", "ID5", "ID6", "ID7", "ID8", "ID9", "ID10"],
-            "View": ["CAUD", "DORS", "FRON", "LATE", "CAUD", "DORS", "FRON", "LATE", "CAUD", "DORS"],
-            "Image": [create_mock_image_blob() for _ in range(10)]
+            "Genus": [
```
Comment (JoeCx) [inline testing/test_genus_specific_model_trainer.py:29]: Data augmentation test fixes

**Review by JoeCx — APPROVED** [review]
Great job adding in hyperparameter tuning for the Genus Specific Training Program. Simulator was updated accordingly.

Additionally, data leakage was fixed in the simulator and moved augmentation into training program.
